HW2/P2/parallel_vector.pyx

```diff
@@ -81,10 +80,15 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads=4):  
                if counts[src[idx]] > 0:
+                   # Add locks and release when done
+                   acquire(&locks[dest[idx]])
                    counts[dest[idx]] += 1
+                   release(&locks[dest[idx]])
+                   acquire(&locks[src[idx]])
                    counts[src[idx]] -= 1
+                   release(&locks[src[idx]])
```

Comment:
agree with @dekatzenel:
Usually, you want to do a check again after acquiring the lock, since the condition might have changed between checking the condition and acquiring the lock